src/Database/DBStructure.php

```diff
@@ -0,0 +1,1689 @@
+		}
+
+		/*if (($parameters["primary"] != "") && $create)
+			$fieldstruct .= " PRIMARY KEY";*/
+
+		return($fieldstruct);
+	}
+
+	private static function createTable($name, $fields, $verbose, $action, $indexes=null) {
+		$r = true;
+
+		$sql_rows = array();
+		$primary_keys = array();
+		foreach ($fields AS $fieldname => $field) {
+			$sql_rows[] = "`".dbesc($fieldname)."` ".self::FieldCommand($field);
+			if (x($field,'primary') && $field['primary']!='') {
+				$primary_keys[] = $fieldname;
+			}
+		}
+
+		if (!is_null($indexes)) {
+			foreach ($indexes AS $indexname => $fieldnames) {
+				$sql_index = self::createIndex($indexname, $fieldnames, "");
+				if (!is_null($sql_index)) {
+					$sql_rows[] = $sql_index;
+				}
+			}
+		}
+
+		$sql = implode(",\n\t", $sql_rows);
+
+		$sql = sprintf("CREATE TABLE IF NOT EXISTS `%s` (\n\t", dbesc($name)).$sql."\n) DEFAULT COLLATE utf8mb4_general_ci";
+		if ($verbose) {
+			echo $sql.";\n";
+		}
+
+		if ($action) {
+			$r = dba::e($sql);
+		}
+
+		return $r;
+	}
+
+	private static function addTableField($fieldname, $parameters) {
+		$sql = sprintf("ADD `%s` %s", dbesc($fieldname), self::FieldCommand($parameters));
+		return($sql);
+	}
+
+	private static function modifyTableField($fieldname, $parameters) {
+		$sql = sprintf("MODIFY `%s` %s", dbesc($fieldname), self::FieldCommand($parameters, false));
+		return($sql);
+	}
+
+	private static function dropIndex($indexname) {
+		$sql = sprintf("DROP INDEX `%s`", dbesc($indexname));
+		return($sql);
+	}
+
+	private static function createIndex($indexname, $fieldnames, $method = "ADD") {
+		$method = strtoupper(trim($method));
+		if ($method!="" && $method!="ADD") {
+			throw new Exception("Invalid parameter 'method' in self::createIndex(): '$method'");
+		}
+
+		if ($fieldnames[0] == "UNIQUE") {
+			array_shift($fieldnames);
+			$method .= ' UNIQUE';
+		}
+
+		$names = "";
+		foreach ($fieldnames AS $fieldname) {
+			if ($names != "") {
+				$names .= ",";
+			}
+
+			if (preg_match('|(.+)\((\d+)\)|', $fieldname, $matches)) {
+				$names .= "`".dbesc($matches[1])."`(".intval($matches[2]).")";
+			} else {
+				$names .= "`".dbesc($fieldname)."`";
+			}
+		}
+
+		if ($indexname == "PRIMARY") {
+			return sprintf("%s PRIMARY KEY(%s)", $method, $names);
+		}
+
+
+		$sql = sprintf("%s INDEX `%s` (%s)", $method, dbesc($indexname), $names);
+		return($sql);
+	}
+
+	private static function groupBy($indexname, $fieldnames) {
+		if ($fieldnames[0] != "UNIQUE") {
+			return "";
+		}
+
+		array_shift($fieldnames);
+
+		$names = "";
+		foreach ($fieldnames AS $fieldname) {
+			if ($names != "") {
+				$names .= ",";
+			}
+
+			if (preg_match('|(.+)\((\d+)\)|', $fieldname, $matches)) {
+				$names .= "`".dbesc($matches[1])."`";
+			} else {
+				$names .= "`".dbesc($fieldname)."`";
+			}
+		}
+
+		$sql = sprintf(" GROUP BY %s", $names);
+		return $sql;
+	}
+
+	public static function definition() {
+		$database = array();
+
+		$database["addon"] = array(
```

Comment:
I'd like to move this to a config file instead. This class shouldn't be modified when we want to change the actual DB structure.

The format of the file is up to you, could be PHP, JSON or even XML or YML (but please don't).

Reply:
As long as you don't want me to change this in this PR, I'm agree.

Reply:
It would be a plus but it isn’t necessary.

Reply:
For this and other stuff (list of emoticons) we need some folder where to store this. Additionally I want to eventually improve this so that addons can contribute database changes as well. But this has to be planned in a longer period.